File: app/src/main/java/com/madebyaron/betterhealth/view/list/ListAdapter.java
package com.madebyaron.betterhealth.view.list;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;
import com.madebyaron.betterhealth.databinding.ItemListBinding;
import com.madebyaron.betterhealth.model.Data;

import java.util.ArrayList;
import java.util.List;

public class ListAdapter extends RecyclerView.Adapter<ListItemHolder> {

    private final FragmentActivity activity;

    private List<Data> dataList = new ArrayList<>();

    private final ItemTouchHelper touchHelper = new ItemTouchHelper(
            new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.START | ItemTouchHelper.END) {

                @Override
                public boolean onMove(
                        @NonNull RecyclerView recyclerView,
                        @NonNull RecyclerView.ViewHolder viewHolder,
                        @NonNull RecyclerView.ViewHolder target
                ) {
                    return false;
                }

                @Override
                public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                    if (viewHolder instanceof ListItemHolder) {
                        ((ListItemHolder) viewHolder).remove();
                    }
                }
            }
    );

    public ListAdapter(FragmentActivity activity) {
        super();
        this.activity = activity;
    }

    public void setData(List<Data> dataList) {
        this.dataList = dataList;
        notifyDataSetChanged();
    }

	public void attachToRecyclerView(RecyclerView recyclerView) {
		recyclerView.setAdapter(this);
		touchHelper.attachToRecyclerView(recyclerView);
	}

    @NonNull
    @Override
    public ListItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return new ListItemHolder(activity, ItemListBinding.inflate(inflater, parent, false));
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    @Override
    public void onBindViewHolder(@NonNull ListItemHolder holder, int position) {
        holder.bind(dataList.get(position));
    }
}

File: app/src/main/java/com/madebyaron/betterhealth/view/list/ListAdapter.kt
package com.madebyaron.betterhealth.view.list

import android.content.ClipData
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.madebyaron.betterhealth.R
import com.madebyaron.betterhealth.data.Data

class ListAdapter(private val activity: FragmentActivity) : RecyclerView.Adapter<ListItemHolder>() {

	var dataList = listOf<Data>()
		set (value) {
			field = value
			notifyDataSetChanged()
		}

	private val touchHelper = ItemTouchHelper(
			object : ItemTouchHelper.SimpleCallback(0,
					ItemTouchHelper.START or ItemTouchHelper.END) {

				override fun onMove(
						recyclerView: RecyclerView,
						viewHolder: RecyclerView.ViewHolder,
						target: RecyclerView.ViewHolder) = false

				override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
					if (viewHolder is ListItemHolder) viewHolder.remove()
				}
			}
	)

	fun attachToRecyclerView(recyclerView: RecyclerView) {
		recyclerView.adapter = this
		touchHelper.attachToRecyclerView(recyclerView)
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListItemHolder {
		val inflater = LayoutInflater.from(parent.context)
		val view = inflater.inflate(R.layout.item_list, parent, false)
		return ListItemHolder(activity, view)
	}

	override fun getItemCount(): Int {
		return dataList.size
	}

	override fun onBindViewHolder(holder: ListItemHolder, position: Int) {
		holder.bind(dataList[position])
	}
}